zebra-state/src/service/non_finalized_state/chain.rs

```diff
@@ -241,6 +253,10 @@ impl UpdateWith<PreparedBlock> for Chain {
             .expect("work has already been validated");
         self.partial_cumulative_work -= block_work;
 
+        // TODO: remove block data from MMR tree
+        // In order to do that, we need to load some "extra" nodes (blocks) from the
+        // state (see zcash_history and librustzcash). How to do that here?
```

Comment:
We might have a bit of a problem here. If we prune the tree every time we add a block, then we'll have to restore those leaves when we remove blocks to fork a chain.

Here's something I think might work:
1. Delete the tree whenever there is a non-finalized fork, but save its nodes as extras
2. Get the finalized tree from the finalized state
3. Rebuild it from the root to the fork using all the blocks in the non-finalized fork, passing the extra nodes
4. Drop the extras once you've reached the fork block, because the next block must be different to the chain you forked off

Since the extras are just an optimisation, we can open a follow-up ticket, add them in a separate PR.

So the initial PR becomes:
1. Delete the tree whenever there is a non-finalized fork
2. Get the finalized tree from the finalized state
3. Rebuild it from the root to the fork using all the blocks in the non-finalized fork

Here are some other alternatives we might be able to implement:
* Only prune the tree when we commit a block to the finalized state. (And if any of the 100 non-finalized blocks are using an entry, we keep it.)

@dconnolly might need to use a similar strategy to handle pruned non-finalized anchor nodes.

Reply:
That seems a good first approach. I've added some logic for it but stumbled upon some issues which I've commented about in the code.

Since most of the nodes/blocks will be in-memory in non-finalized chains, I believe that another option would be to compute the extra nodes from the in-memory blocks and try to remove the node(s). If it fails (because it tries to load a node from the finalized state) it could revert to this approach of rebuilding the tree from the finalized state.

Reply:
I think there are a few alternative optimisations here, and I'm happy to try them.

But let's write the simplest possible working code first, and use it to write tests. Then our tests can help us find out which optimisations actually work.